### DISCUSSION TITLE
ProxyJump not working on 17.0.0

### DISCUSSION THREAD

**Discussion by tjim** [blinksh/blink#1868] [Support] [answered]
### Checklist

- [x] Using latest version on the [App Store](http://itunes.apple.com/app/id1156707581)
- [x] Read the [docs](http://blink.sh/docs/)
- [x] Searched for [existing GitHub issues](https://github.com/blinksh/blink/issues)

### Configuration

Just updated to blink 17.0.0, iPad OS 16.1.1

### Describe the bug

I have configured host XXXXX to use host YYYYY as ProxyJump. In the previous version of blink I was able to

    ssh XXXXX

In the new version 17.0.0, this hangs. When I ssh -vvvvv XXXXX, the debug message just before the hang is

    ssh_socket_connect_proxycommand: Executing proxycommand 'ssh -W [XXXXX]:22 YYYYY'

I am able to log in to YYYYY fine with ‘ssh YYYYY’.

I am able to log in to XXXXX using YYYYY as a jump host in iSH on the same iPad (ssh -J YYYYY XXXXX).

Possibly related: ssh-agent is not working.

    blink> ssh-agent
    Failed loading ssh_agent_main from MAIN, cause = dlsym(RTLD_MAIN_ONLY, ssh_agent_main): symbol not found
    ssh-agent: command not found

**Comment by carloscabanero** [comment]
These are usually configuration issues. Could you send the .blink/ssh_config file? And also, the full output from -vvvvv? I will try to replicate from my side.

Previously discussed here https://github.com/blinksh/blink/discussions/1840

**Reply by carloscabanero** [reply]
Oh, forgot to mention. The issue with ssh-agent is a rare issue with ios_system "executing" binaries in Blink. Usually restarting the app should do the trick. That may be what is causing the issue itself as there is another process in the background for the ProxyCommand, so worth trying.

**Reply by tjim** [reply]
I restarted the app, did not fix ssh-agent issue. Rebooted the iPad, this also did not fix ssh-agent (results in the same behavior as above).

**Reply by carloscabanero** [reply]
Sorry for the confusion. I thought you meant ssh-add. ssh-agent runs by default and is disabled, you don’t need to execute it yourself. Are you adding keys to the agent for log in through the proxy? I would be more able able to help you with the information from ssh -vvvvv and the blink/ssh_config file. If sending here is a problem, please do it to support at blink.sh

**Comment by tjim** [comment]
OK, here is config, I have had to rename some hosts to protect the innocent:

    # ATTENTION! THIS IS GENERATED FILE. DO NOT CHANGE IT DIRECTLY.
    # GENERATED 2023-09-20T17:29:43Z
    #
    # Use config command do configure your hosts
    # Or put your configuration to ~/.ssh/config

    Host YYYYY
      User tjim
      HostName BBBBB
      IdentityFile id_rsa
      ForwardAgent yes

    Host XXXXX
      User tjim
      HostName AAAAA
      IdentityFile id_rsa
      ProxyJump YYYYY

Full output from -vvvvv:

    blink> ssh -vvvvv XXXXX
    Setting up connection callbacks.
    Starting connection to AAAAA
    ssh_connect: libssh 0.9.3 (c) 2003-2019 Aris Adamantiadis, Andreas Schneider and libssh contributors. Distributed under the LGPL, please refer to COPYING file for information about your rights, using threading threads_pthread
    ssh_socket_connect_proxycommand: Executing proxycommand 'ssh -W [AAAAA]:22 YYYYY'

At that point it hangs.

The ssh-agent problem only came up because I was trying to debug, I don’t usually invoke it. (I was trying to log in manually to the jump host YYYYY, then ssh from there to XXXXX.) In other words, I get the behavior above if ssh XXXXX is the first command I issue to blink on startup.

**Comment by carloscabanero** [comment]
Ok, I will try and reproduce it. One more question from my side.

If you first connect to YYYYY and then in another tab connect to XXXXX, does that go through?

Weird that this broke on iOS17 for you, I think there must be something very strange going on.

**Comment by tjim** [comment]
If I first connect to YYYYY and then in another tab connect to XXXXX, it DOES NOT go through.

Sounds like ssh-agent is a red herring. I suggest that you remove it from completions (that is, when I type ‘ssh-a’ it shows as one of the possible completions), otherwise it may be confusing for someone used to Linux.

If I ssh-add, then ssh YYYYY, then I can ssh XXXXX.

I don’t know if there is something strange on my system; I will say that ProxyJump did not work for me until 16.0, but it seemed to work for a lot of other people before that. Previously I had to ssh-add, ssh YYYYY, then ssh XXXXX, which is what is working for me now.

**Reply by carloscabanero** [reply]
Ok, thanks for the detailed debugging. I will consider this blocking so I am jumping on it now to see if I can replicate it and figure out what may be going on.

I don't see why in your case you need the ssh agent though (it is one of those cases where the -W flag tunnel replaced the whole bastion server forwarding connections). You would need the agent if from XXXXX you wanna connect to, for example, GitHub, using one of the keys in your local iPad. That is because in that case XXXXX would be the owner of the ssh connection to the remote. But to connect to XXXXX and YYYYY, all the authentication is still managed by your iPad as the connection to the Proxy and the destination are still owned by the iPad.

**Reply by tjim** [reply]
Agreed, in my normal use with blink 16.x, I never used ssh-agent or ssh-add, I just used the config above, with ProxyJump.

In debugging I was trying things that I had needed in previous versions of blink where ProxyJump was not working for me (in some early versions of blink, I believe ssh-agent was necessary, and since I was manually going to YYYYY and from there to XXXXX, I needed -A and therefore ssh-add).

**Reply by carloscabanero** [reply]
Ok, found what is going on. I was banging my head because I could replicate but I could not debug when installing from Xcode. The issue is that when we shipped, the build used a different library last minute looks like that is what is breaking this. 

Are you on TestFlight by any chance? If you roll back to Build 833, things should work fine there. Let me know if not, I can give you access to confirm from your side.

In the meantime, preparing another 17.0.1 build to fix this.

**Reply by tjim** [reply]
I have TestFlight but have been using the regular version of blink. If you’d like me to confirm I’d need to be set up on TestFlight. I can also wait for 17.0.1, whatever works.

Thank you for fixing!

**Reply by carloscabanero** [reply]
Just released 17.0.1. Let us know

**Reply by tjim** [reply]
Behavior is different but still unable to connect.

v17.0.1.840

    blink> ssh -vvvvv XXXXX
    Setting up connection callbacks.
    Starting connection to AAAAA
    ssh_connect: libssh 0.9.3 (c) 2003-2019 Aris Adamantiadis, Andreas Schneider and libssh contributors. Distributed under the LGPL, please refer to COPYING file for information about your rights, using threading threads_pthread
    ssh_socket_connect_proxycommand: Executing proxycommand 'ssh -W [AAAAA]:22 YYYYY'
    ssh_socket_connect_proxycommand: ProxyCommand connection pipe: [58,59]
    socket_callback_connected: Socket connection callback: 1 (0)
    ssh_connect: Socket connecting, now waiting for the callbacks to work
    ssh_connect: current state : 2
    Starting connection to AAAAA
    ssh_connect: current state : 2
    Starting connection to AAAAA
    ssh_connect: current state : 2
    ssh_socket_exception_callback: Socket exception callback: 1 (0)
    ssh_socket_exception_callback: Socket error: disconnected
    connError(msg: "Session Exception - Socket error: disconnected")
    Starting connection to AAAAA
    ssh_connect: current state : 9
    Error connecting to XXXXX. connError(msg: "Socket error: disconnected")
    Session deinit

As before, I am able to ssh-add, ssh YYYYY, and from there ssh XXXXX, it is just ProxyJump that is not working.

Also, it still does not work if I am logged in to YYYYY in another tab.

I have seen this kind of behavior on blink versions < 16, when it seemed like ProxyJump was working for others but not me.

**Reply by carloscabanero** [reply]
Yeah, but we can see more now. The other side is disconnecting when we do the stdio (-W) tunnel. We haven’t touched this from 16 to 17 so it is weird that this worked for you before. We can figure it out though, just need you to run a couple quick tests for me, if you still have the patience:

- In a different shell, run 'ssh -W [AAAAA]:22 YYYYY'. It should stay connected and blank. If it disconnects, then add -vvvvv so we can see what is going on.
- Does the Files.app integration work? We use a different method to do the jump and that would point me on a direction for the fix.

Thank you!

**Reply by tjim** [reply, marked as answer]
This is working now! I did a reboot, that may have done something. Files.app also works.
Thanks!

**Reply by carloscabanero** [reply]
Ok, I will keep an eye on this one in case it repeats. I understand you did a reboot of the device and not the server, right?

**Comment by yury** [comment]
Tracked down original issue and made [PR](https://github.com/holzschu/ios_system/pull/148)
